File: src/Systems/Zig/iterator.zig
const std = @import("std");

const Iterator = struct {
    values: []const i32,
    index: usize = 0,

    fn next(self: *Iterator) ?i32 {
        if (self.index >= self.values.len) return null;
        const value = self.values[self.index];
        self.index += 1;
        return value;
    }
};

pub fn runIteratorExample() bool {
    const values = [_]i32{ 10, 20, 30 };
    var iterator = Iterator{ .values = &values };
    var visited = [_]i32{ 0, 0, 0 };
    var index: usize = 0;

    while (iterator.next()) |value| {
        visited[index] = value;
        index += 1;
    }

    return index == 3 and visited[0] == 10 and visited[1] == 20 and visited[2] == 30 and iterator.next() == null;
}

pub fn main() !void {
    if (!runIteratorExample()) return error.IteratorContractFailed;
    std.debug.print("iterator=10,20,30\n", .{});
}

File: src/Systems/Go/iterator.go
package main

import "fmt"

type Iterator[T any] struct {
	values []T
	index  int
}

func (it *Iterator[T]) Next() (T, bool) {
	if it.index >= len(it.values) {
		var zero T
		return zero, false
	}
	value := it.values[it.index]
	it.index++
	return value, true
}

func main() {
	it := Iterator[int]{values: []int{10, 20, 30}}
	visited := make([]int, 0, 3)
	for {
		value, ok := it.Next()
		if !ok {
			break
		}
		visited = append(visited, value)
	}
	if len(visited) != 3 || visited[0] != 10 || visited[1] != 20 || visited[2] != 30 {
		panic("iterator contract failed")
	}
	fmt.Println("iterator=10,20,30")
}
